eidr/index.go

```diff
@@ -0,0 +1,420 @@
+				return err
+			}
+			if err := i.index(obj); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+// Based on EIDR 2.0 Data Types
+func (i *Indexer) index(eidrobj *meta.Object) error {
+
+	topgraph := meta.NewGraph(i.store, eidrobj)
+
+	// eidr objects are divided in baseobject and extraobject
+	// we extract these top level objects first
+	base, err := topgraph.Get("FullMetadata", "BaseObjectData")
+	if meta.IsPathNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	var extracid *cid.Cid
+	var extraobj *meta.Object
+	extra, err := topgraph.Get("FullMetadata", "ExtraObjectMetadata")
+	if err != nil && !meta.IsPathNotFound(err) {
+		return err
+	} else {
+		var ok bool
+		extracid, ok = extra.(*cid.Cid)
+		if !ok {
+			return nil
+		}
+		parselog.Log(log.LL_DEBUG, "eidr", fmt.Sprintf("extradata cid %v", *extracid))
+		extraobj, err = i.store.Get(extracid)
+		if err != nil {
+			return err
+		}
+	}
+	basecid, ok := base.(*cid.Cid)
+	if !ok {
+		return nil
+	}
+	baseobj, err := i.store.Get(basecid)
+	if err != nil {
+		return err
+	}
+
+	// PARSE BASE OBJECT (common for all types)
+
+	var baseObject struct {
+		simple            map[string]string
+		id                doi.ID
+		status            string
+		structuralType    string
+		associatedOrg     []AssociatedOrg
+		referentType      string
+		resourceName      string
+		resourceNameClass *string
+		resourceNameLang  string
+		alternateID       []AlternateID
+		approximateLength *string
+		parentId          *doi.ID
+	}
+	// simplified handling of key/value pairs
+	baseObject.simple = make(map[string]string)
+
+	// iterate all elements
+	var m map[string]interface{}
+	baseobj.Decode(&m)
+	for field, _ := range m {
+		parselog.Log(log.LL_WARN, basecid.String(), field)
+		l, err := baseobj.GetLink(field)
+		if err != nil {
+			// don't fail on optional fields: alternateID
+			// pass @value through as we use it for the simple handler
+			if field == "AlternateID" || (field[0:1] == "@" && field != "@value") {
+				continue
+			}
+			fmt.Printf("linknotfound: %v", err)
+			return err
+		}
+		// get current object and graph
+		lobj, err := i.store.Get(l.Cid)
+		if err != nil {
+			return err
+		}
+		graph := meta.NewGraph(i.store, lobj)
+		toptype, err := graph.Get("@type")
+		if err != nil {
+			return err
+		}
+		v, err := graph.Get("@value")
+		switch toptype.(string) {
+		case "Credits":
+			// todo
+			break
+		case "Administrators":
+			// todo
+			break
+		case "ID":
+			baseObject.id = doi.ID(v.(string))
+			break
+		case "ResourceName":
+			lang, err := lobj.GetString("lang")
+			if err != nil {
+				return err
+			}
+			class, _ := lobj.GetString("titleClass") // optional
+			baseObject.resourceName = v.(string)
+			baseObject.resourceNameLang = lang
+			baseObject.resourceNameClass = &class
+			break
+		case "AlternateID":
+			t, err := graph.Get("http://www.w3.org/2001/XMLSchema-instance:type")
```

Comment:
Did you figure out why the full URL is in the property name? Seems like a bug.

Reply:
I didn't look into it. I'll make an issue of it and investigate.